_posts/blog/2016-09-06-model-view-viewmodel.md

```diff
@@ -0,0 +1,181 @@
+                         .observeOn(AndroidSchedulers.mainThread())
+                         .subscribe(this::updateView,
+                                    this::handleError));
+    }
+
+    @Override
+    public void onDetachedFromWindow() {
+        mSubscription.clear();
+        super.onDetachedFromWindow();
+    }
+}
+{% endhighlight %}
+
+## Testability Of The Model-View-ViewModel Classes
+
+One of the main reasons why we love the Model-View-ViewModel pattern is its high degree of testability.
+
+### DataModel
+
+The inversion of control pattern, heavily applied in our code, and the lack of any Android classes, make the DataModel easy to unit test.
+
+### ViewModel
+
+The ViewModel exposes streams of data to whoever needs to consume it. We see the Views and the unit tests as two different types of data consumers.
+The ViewModel is completely separated from the UI or any Android classes, therefore straightforward to unit test.
+
+Considering the following example where the ViewModel just exposes some data from the DataModel:
+
+{% highlight java %}
+public class ViewModel {
+    @NonNull
+    private final IDataModel mDataModel;
+
+    public ViewModel(@NonNull final IDataModel dataModel) {
+        mDataModel = dataModel;
+    }
+
+    @NonNull
+    public Observable<Data> getSomeData() {
+        return mDataModel.getSomeData();
+    }
+}
+{% endhighlight %}
+
+The tests for the ViewModel are easy to implement. With the help of Mockito, we are mocking the DataModel and we control the returned data for the methods used. Then, we make sure that when we subscribe to the `Observer` returned by `getSomeData()`, the expected data is emitted.  
+
+{% highlight java %}
+public class ViewModelTest {
+
+    @Mock
+    private IDataModel mDataModel;
+
+    private ViewModel mViewModel;
+
+    @Before
+    public void setUp() throws Exception {
+        MockitoAnnotations.initMocks(this);
+
+        mViewModel = new ViewModel(mDataModel);
+    }
+
+    @Test
+    public void testGetSomeData_emitsCorrectData() {
+        SomeData data = new SomeData;
+        Mockito.when(mDataModel.getSomeData()).thenReturn(Observable.just(data));
+        TestSubscriber<SomeData> testSubscriber = new TestSubscriber<>();
+
+        mViewModel.getSomeData().subscribe(testSubscriber);
+
+        testSubscriber.assertValue(data);
+    }
+}
+{% endhighlight %}
+
+If the ViewModel needs access to Android classes, we create wrappers that we call `Provider`s. For example, for resources we created a `IResourceProvider`, that exposes methods like `String getString(@StringRes final int id)`. The implementation of the `IResourceProvider` will contain a reference to the `Context` but, the ViewModel will only get a reference to a `IResourceProvider` injected.
+
+### View
+
+Given that the logic in the UI is minimal, the Views are easy to test with Espresso. We are also using libraries like DaggerMock and MockWebServer to improve the stability of our UI tests.
```

Comment:
Repetition, and it's probably not so much that are *easy* to test, but more like there's not much to test in terms of possible states, control flow point etc (logic). 
> easy to test